### PR TITLE
Build, test and publish wheels via Github Action, remove cmake

### PR DESCRIPTION
**Pull request by chaitan94**
## Wheels building status

1. ✅ **Linux**: wheels seem to build for manylinux without any issues.
2. ❌ **Mac OS**: failing with a project level error:
```
   [ 50%] Building CXX object CMakeFiles/pytomlpp.dir/src/pytomlpp.cpp.o
  In file included from /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-req-build-ojfeoq17/src/pytomlpp.cpp:3:
  /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-req-build-ojfeoq17/include/pytomlpp.hpp:38:99: error: 'value' is unavailable: introduced in macOS 10.14
          py::object time_delta = PY_DATETIME_MODULE.attr("timedelta")("minutes"_a = dt.time_offset.value().minutes);
                                                                                                    ^
  /Applications/Xcode_11.5.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/optional:943:33: note: 'value' has been explicitly marked unavailable here
      constexpr value_type const& value() const&
                                  ^
  1 error generated.
```
3. ❓ **Windows**: Definitely the trickiest. Have to test once Mac OS support is in place.

# Publishing workflow
Note: We do not build and publish sdist to pip. Ideal publishing flow for any new release should be:

(Assuming we are publishing a version say 1.2.3)

## Publish to test PyPI

1. Set file content of VERSION to 1.2.3rc1
2. Build sdist locally: `python setup.py sdist`
3. Publish sdist: `twine upload --repository testpypi dist/*`
4. Test install works locally `pip install -i https://test.pypi.org/simple/ pytomlpp==1.2.3rc1`
5. If something fails, go to step 1, increment version, say rc2, repeat until success

## Publish to PyPI
(First make sure PYPI_TOKEN is setup in your repository's Github secrets)

1. Set file content of VERSION to 1.2.3
2. Add a new commit `Bump to 1.2.3`
3. Push commit `git push`
4. Verify in GitHub Actions that all wheels are built as expected (note they won't be published yet, as we didn't tag the commit)
5. If any of them fail, you should fix the issues accordingly before proceeding further
6. Tag your commit `git tag 1.2.3`
7. Push commit `git push origin 1.2.3`
8. Sit back and relax as your wheels get built and released on PyPI

### CODE REVIEW

**Comment by zooba** [comment]
Windows should be trivial for this, I don't understand what's so tricky?

The GitHub Actions images include CMake and Visual Studio for Python 3.5 and later, so once you drop 2.7 support it ought to be straightforward.

Do some of the build tools you're using need contributions?

**Comment by EpicWink** [comment]
I would suggest you pull in the changes of #2, That would remove the need for Cmake, especially as then we're testing just the library itself. This could also fix the issues with Windows

**Comment by bobfang1992** [comment]
Looking at this, and I think I agree with @EpicWink on where this should go: we should remove cmake altogether and just use python setup.py/pip to build this package. Also can remove the google tests as there is no need for them. @chaitan94 Do you want to make this change? Thanks!

**Comment by bobfang1992** [comment]
I guess to do is to:

- [ ] remove pybind11 and google test submodule
- [ ] remove C++ unit tests 
- [ ] remove CMakeLists.txt file

And then we are in good shape.

**Comment by chaitan94** [comment]
Sure will make suggested changes. 

Apart from that are any of you aware of what could be the issue I'm facing in Mac OS build? I am not familiar with C++17 features as much yet. Seems like `optional` feature is not working as expected in mac.

**Comment by bobfang1992** [comment]
@chaitan94 I will look into that and see if I can help you.

**Comment by bobfang1992** [comment]
@chaitan94 I merged #3, please rebase, thanks.

**Comment by chaitan94** [comment]
Regarding google tests, I guess we wouldn't need `test_lib_config.cpp` as well right?

**Comment by bobfang1992** [comment]
> Regarding google tests, I guess we wouldn't need `test_lib_config.cpp` as well right?

yep, no need for it anymore @chaitan94

**Comment by chaitan94** [comment]
I've pulled in all latest changes and removed google-tests related code. Right now wheels are being built both by Github Actions and Travis (Although we have yet to fix the macOS issue on Github Actions). Should we deprecate Travis?

**Comment by bobfang1992** [comment]
Yeah, let's deprecate Travis. Give me some time and I will see if I can fix the Mac OS issue. BTW is windows okay now or you need to work on it?

File: .gitmodules [hunk 1]
```diff
@@ -1,9 +0,0 @@
-[submodule "pybind11"]
-	path = pybind11
-	url = https://github.com/pybind/pybind11.git
-[submodule "googletest"]
-	path = googletest
-	url = https://github.com/google/googletest.git
-[submodule "tests/toml-test"]
```
Comment (bobfang1992) [inline .gitmodules:7]: Why remove `toml-test`? The python unit tests need this module to work.

**Comment by chaitan94** [comment]
Right now its failing with 'unknown option '-std=c++17', I think MSVC expects something like '/std:c++17' instead of '-std=c++17'. I will try that and check accordingly.

File: .gitmodules (hunk 1 above)
Comment (chaitan94) [reply .gitmodules:7]: Oops sorry will add that back.

**Comment by chaitan94** [comment]
PyPy build also seems to fail (even on Linux): https://github.com/chaitan94/pytomlpp/runs/790601180?check_suite_focus=true
```
   In file included from src/pytomlpp.cpp:3:
  include/pytomlpp.hpp: In function ‘toml::array pytomlpp::py_list_to_toml_array(const pybind11::list&)’:
  include/pytomlpp.hpp:264:77: error: cannot bind non-const lvalue reference of type ‘pybind11::handle&’ to an rvalue of type ‘pybind11::handle’
               toml::date_time date_time_value = py_datetime_to_toml_date_time(it);
                                                                               ^~
```

**Comment by chaitan94** [comment]
I have added back toml-test. I also made a change so that tests won't succeed without it.

Guys I have restricted to Linux and non-pypy. I think it would be best if we merge the work so far and continue work on fixing Mac OS, Windows and PyPy build seperately. Since there are many commits already constantly rebasing would be a inefficient use of time.

**Comment by chaitan94** [comment]
The build and test is succeeding as of now: https://github.com/chaitan94/pytomlpp/runs/790658356?check_suite_focus=true

**Comment by bobfang1992** [comment]
Hi, Thanks for all the hard work. So the gist is:

- This change deprecates Travis and uses Github Actions instead.
- it builds on Linux 
- it does not build on Windows or Mac OS yet, also not working for Pypy

Is this understanding correct? I will be happy to merge this if it is. And then we can work on separate PRs to get the above limitations removed.

Just 2 points:

1. Let me do a quick review
2. Could you rebase? Seems currently we have conflicts.

Thanks!

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,23 +1,90 @@
-cmake_minimum_required(VERSION 3.9.0)
```
Comment (bobfang1992) [inline CMakeLists.txt:1]: Do we still need this file?

File: cmake/CPM.cmake [hunk 3]
```diff
@@ -0,0 +1,475 @@
+# CPM.cmake - CMake's missing package manager
```
Comment (bobfang1992) [inline cmake/CPM.cmake:1]: Also do we need this one?

File: .github/workflows/wheels-linux.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        submodules: true
+
+    - uses: actions/setup-python@v2
+      name: Install Python
+      with:
+        python-version: 3.8
```
Comment (bobfang1992) [inline .github/workflows/wheels-linux.yml:22]: Hi, the current Travis CI builds on 3.5, 3.6, 3.7, 3.8 and 3.9. Can Github Actions do the same? Thanks!

File: .github/workflows/wheels-linux.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    - name: Build wheels
+      env:
+        # Skip building on Python 2.7 and PyPy
+        CIBW_SKIP: cp27-* pp*
+        CIBW_BEFORE_TEST: pip install -r tests/requirements.txt
+        CIBW_TEST_COMMAND: pytest {project}/tests
+      run: |
+        python -m cibuildwheel --output-dir dist
+
+    - uses: actions/upload-artifact@v2
```
Comment (bobfang1992) [inline .github/workflows/wheels-linux.yml:37]: Hi where does this upload to?

File: .github/workflows/wheels-linux.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  publish:
+    runs-on: ubuntu-latest
+    needs: [build]
+    if: github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags/')
+    steps:
+      - uses: actions/download-artifact@v2
+        with:
+          name: dist
+          path: dist/
+      - name: Publish to PyPI
```
Comment (bobfang1992) [inline .github/workflows/wheels-linux.yml:51]: If we want to leverage this later to publish to pypi, where do I set the `__token__` etc? Do you have any documentation on this?
Thanks!

File: .github/workflows/wheels-linux.yml (hunk 4 above)
Comment (chaitan94) [reply .github/workflows/wheels-linux.yml:22]: This is just the base image's python version. Actual wheels are built for 3.5, 3.6, 3.7, 3.8 transparently by cibuildwheel

Refer: https://cibuildwheel.readthedocs.io/en/stable/options/#build-skip

File: .github/workflows/wheels-linux.yml (hunk 5 above)
Comment (chaitan94) [reply .github/workflows/wheels-linux.yml:37]: Github stores in temporary storage for the CI run. It is private to that particular run and gets cleaned later on. Specifically, in this run, we will later collect all these wheels in the publish step (which is only run when you push a git tag). If you setup PYPI_TOKEN in GitHub secrets, you can automatically publish to PyPI when pushing a new git tag.

File: .github/workflows/wheels-linux.yml (hunk 6 above)
Comment (chaitan94) [reply .github/workflows/wheels-linux.yml:51]: You should be able to add here: https://github.com/bobfang1992/pytomlpp/settings/secrets

Get the token from your PyPI account, you can create one just for the repo ideally.

File: CMakeLists.txt (hunk 2 above)
Comment (chaitan94) [reply CMakeLists.txt:1]: You are right, I will remove all cmake related stuff and push a commit.

File: cmake/CPM.cmake (hunk 3 above)
Comment (chaitan94) [reply cmake/CPM.cmake:1]: You are right, I will remove all cmake related stuff and push a commit.

**Comment by chaitan94** [comment]
> Hi, Thanks for all the hard work. So the gist is:
> 
>     * This change deprecates Travis and uses Github Actions instead.
> 
>     * it builds on Linux
> 
>     * it does not build on Windows or Mac OS yet, also not working for Pypy
> 
> 
> Is this understanding correct? I will be happy to merge this if it is. And then we can work on separate PRs to get the above limitations removed.
> 
> Just 2 points:
> 
>     1. Let me do a quick review
> 
>     2. Could you rebase? Seems currently we have conflicts.
> 
> 
> Thanks!

Yes, you are correct on all three points. I have resolved the conflicts just a few minutes ago - there shouldn't be any now.
I have remove cmake realted stuff. We should be good to merge once this suceeds: https://github.com/chaitan94/pytomlpp/actions/runs/141790648 .

File: .github/workflows/wheels-linux.yml (hunk 6 above)
Comment (bobfang1992) [reply .github/workflows/wheels-linux.yml:51]: Great

File: CMakeLists.txt (hunk 2 above)
Comment (bobfang1992) [reply CMakeLists.txt:1]: Thanks!

File: cmake/CPM.cmake (hunk 3 above)
Comment (bobfang1992) [reply cmake/CPM.cmake]: Thanks!

File: .github/workflows/wheels-linux.yml (hunk 5 above)
Comment (bobfang1992) [reply .github/workflows/wheels-linux.yml:37]: Good thanks for the clarification.

File: .github/workflows/wheels-linux.yml (hunk 6 above)
Comment (bobfang1992) [reply .github/workflows/wheels-linux.yml:51]: Good thanks for the clarification.

Comment (EpicWink) [reply .github/workflows/wheels-linux.yml:51]: Action documentation: https://github.com/pypa/gh-action-pypi-publish

**Comment by chaitan94** [comment]
> Windows should be trivial for this, I don't understand what's so tricky?

@zooba Windows is ALWAYS the trickiest. Never underestimate it :stuck_out_tongue_closed_eyes:.

Hence, let's keep it and macos support separate from this for now.

**Comment by bobfang1992** [comment]
Hi, sorry I still see there is conflicts and cannot merge... Can you check again? @chaitan94

**Comment by bobfang1992** [comment]
> This branch cannot be rebased due to conflicts
Rebasing the commits of this branch on top of the base branch cannot be performed automatically due to conflicts encountered while reapplying the individual commits from the head branch.

This is what I am seeing right now.

**Comment by chaitan94** [comment]
Maybe you want to squash and merge instead?

File: tests/python-tests/test_api.py [hunk 7]
```diff
@@ -27,8 +27,10 @@
 ]
 
 _toml_dir = pathlib.Path(__file__).parent.parent / "toml-test" / "tests"
-valid_toml_files = (_toml_dir / "valid").glob("*.toml")
-invalid_toml_files = (_toml_dir / "invalid").glob("*.toml")
+valid_toml_files = list((_toml_dir / "valid").glob("*.toml"))
+invalid_toml_files = list((_toml_dir / "invalid").glob("*.toml"))
+if len(valid_toml_files) <= 0 or len(invalid_toml_files) <= 0:
```
Comment (EpicWink) [inline tests/python-tests/test_api.py:32]: `len(list) <= 0` 🤣

**Comment by EpicWink** [comment]
You can do this as well
```bash
git fetch -p --all
git reset --soft upstream/master
git add -u
git commit
```

**Comment by chaitan94** [comment]
> This branch cannot be rebased due to conflicts

@bobfang1992 I made a clean rebase again now. Hopefully, you shouldn't face any issues now.

**Comment by EpicWink** [comment]
Ahh yeah, `pathlib.Path.glob` is a generator. You'll need to `valid_toml_files = list(valid_toml_files)`

**Comment by bobfang1992** [comment]
Great thanks! This is a lot of work. Thank you a lot @chaitan94

**Pull request merged**